metricbeat/internal/metrics/memory/memory.go

```diff
@@ -0,0 +1,124 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package memory
+
+import (
+	"github.com/pkg/errors"
+
+	"github.com/elastic/beats/v7/libbeat/common"
+	"github.com/elastic/beats/v7/metricbeat/internal/metrics"
+	"github.com/elastic/go-structform/gotype"
+)
+
+// Memory holds os-specifc memory usage data
+// The vast majority of these values are cross-platform
+// However, we're wrapping all them for the sake of safety, and for the more variable swap metrics
+type Memory struct {
```

Comment:
I wonder if we really should have a common type for all systems with optionals, or rather want a custom type per system without optionals.

Reply:
By "system" do you mean OS/platform? That's actually what I did fairly early on in the refactor process, starting with `system/cpu`, but creating different per-system types didn't seem worth it, as the overwhelming majority of the metrics were cross-platform, and there were drawbacks to getting clever with interfaces and so on when you have multiple types like that. I ended up abandoning it in favor of a universal struct for all platforms.

Reply:
Sounds fair.